### PR TITLE
sql: unit test that runs a multi-node cluster

### PR DESCRIPTION
**Pull request by dt**
Just a helper to setup/tear down multi-node cluster in a single process and a simple unit test that it works as expected. Groundwork for automated benchmarks of multi-node perf.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4775)

### CODE REVIEW

File: sql/multinode_test.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		}
+		for _, fn := range cleanups {
+			fn()
+		}
+	}
+
+	return conns, f
+}
+
+func TestMultinodeCockroack(t *testing.T) {
```
Comment (tamird) [inline sql/multinode_test.go]: Cockroack?

File: server/testserver.go [hunk 2]
```diff
@@ -63,6 +63,22 @@ func StartTestServer(t util.Tester) *TestServer {
 	return s
 }
 
+// StartTestServerJoining starts an in-memory test server that attempts to join `other`.
+func StartTestServerJoining(t util.Tester, other *TestServer) *TestServer {
```
Comment (tamird) [inline server/testserver.go]: this doesn't use any unexported fields, is there any reason it needs to be linked into the production binary?

**Comment by dt** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[server/testserver.go, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4775#-KBsQL71INIiWEqLx8tQ-r1-67:-KBs_kbt1sC79hGkyo3c:-1113919877) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4ce04a9864c30ce7b1b8e884441feb97a496a45/server/testserver.go#L67)):*
`TestServer` definitely uses some of the unexpected fields of Server, so it has to live in `server`. The other constructors for `TestServer`s are here, so I don't know that it makes sense to just move this one.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4775)*

**Comment by dt** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[sql/multinode_test.go, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4775#-KBsQL71INIiWEqLx8tR-r1-74:-KBsaA1UK1v6yhbVkw4X:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4ce04a9864c30ce7b1b8e884441feb97a496a45/sql/multinode_test.go#L74)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4775)*

File: server/testserver.go [hunk 3]
```diff
@@ -63,6 +63,22 @@ func StartTestServer(t util.Tester) *TestServer {
...
+	s := &TestServer{Ctx: NewTestContext()}
+	a := other.ServingAddr()
```
Comment (nvanbenschoten) [inline server/testserver.go]: Any reason for this to be split into a different variable?

Comment (dt) [inline server/testserver.go]: oh, yeah, I split it out when it was panicking to get a more useful line number. I can put it back inline.

File: sql/multinode_test.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	"testing"
+
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/server"
+	"github.com/cockroachdb/cockroach/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/util/leaktest"
+	"github.com/cockroachdb/cockroach/util/tracing"
+)
+
+func SetupMultinodeTestCluster(t testing.TB, nodes int, name string) ([]*sql.DB, func()) {
```
Comment (nvanbenschoten) [inline sql/multinode_test.go]: You might want to assert that nodes is >= 1

Comment (nvanbenschoten) [inline sql/multinode_test.go]: You also should probably add some comment to this function explaining what it does and what it returns. Specifically it took looking into the implementation to see what the returned function is, so anything that clears that up should be fine.

**Comment by nvanbenschoten** [comment]
LGTM, really glad to see this!

**Comment by dt** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[server/testserver.go, line 69 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4775#-KBsQL71INIiWEqLx8tQ-r2-69:-KBshEAG8jSR1byQZcv7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1130c73961eadf6cfa47a3aab205349db2580c65/server/testserver.go#L69)):*
Done.

---

*[sql/multinode_test.go, line 31 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4775#-KBsQL71INIiWEqLx8tR-r2-31:-KBshEuDVb-9ul-3d5Jl:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1130c73961eadf6cfa47a3aab205349db2580c65/sql/multinode_test.go#L31)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4775)*

**Pull request merged**